milli/src/search/facet/filter.rs

```diff
@@ -89,52 +89,44 @@ impl<'a> Filter<'a> {
         I: IntoIterator<Item = Either<J, &'a str>>,
         J: IntoIterator<Item = &'a str>,
     {
-        let mut ands: Option<FilterCondition> = None;
+        let mut ands = vec![];
 
         for either in array {
             match either {
                 Either::Left(array) => {
-                    let mut ors = None;
+                    let mut ors = vec![];
                     for rule in array {
                         if let Some(filter) = Self::from_str(rule.as_ref())? {
-                            let condition = filter.condition;
-                            ors = match ors.take() {
-                                Some(ors) => {
-                                    Some(FilterCondition::Or(Box::new(ors), Box::new(condition)))
-                                }
-                                None => Some(condition),
-                            };
+                            ors.push(filter.condition);
                         }
                     }
 
-                    if let Some(rule) = ors {
-                        ands = match ands.take() {
-                            Some(ands) => {
-                                Some(FilterCondition::And(Box::new(ands), Box::new(rule)))
-                            }
-                            None => Some(rule),
-                        };
+                    if ors.len() > 1 {
+                        ands.push(FilterCondition::Or(ors));
+                    } else if ors.len() == 1 {
+                        ands.push(ors[0].clone());
```

Comment:
```suggestion
                        ands.push(ors[0].pop().unwrap());
```

Shouldn't this work?